File: FRONTEND/src/service/CommentService.ts
import apiClient from "@/config/ApiClient";
import type { Comment } from "@/models/Comment";

// Video ke saare comments lo (public — login zaroori nahi)
export const getComments = async (videoId: string): Promise<Comment[]> => {
    const response = await apiClient.get<Comment[]>(`/videos/${videoId}/comments`);
    return response.data;
};

// Naya comment post karo (login zaroori)
// parentCommentId diya to ye kisi comment ka reply banega
export const postComment = async (
    videoId: string,
    text: string,
    parentCommentId?: string
): Promise<Comment> => {
    const response = await apiClient.post<Comment>(`/videos/${videoId}/comments`, {
        text,
        parentCommentId: parentCommentId ?? null,
    });
    return response.data;
};

// Comment delete karo — sirf apna, ya admin kisi ka bhi
export const deleteComment = async (commentId: string): Promise<void> => {
    await apiClient.delete(`/comments/${commentId}`);
};

// ── ADMIN ────────────────────────────────────────────────────────────────

// Admin: poore platform ke saare comments lo (moderation ke liye)
export const getAllCommentsAdmin = async (): Promise<Comment[]> => {
    const response = await apiClient.get<Comment[]>("/admin/comments");
    return response.data;
};

// Admin: kisi bhi comment ko force delete karo
export const deleteCommentAdmin = async (commentId: string): Promise<void> => {
    await apiClient.delete(`/admin/comments/${commentId}`);
};

File: BACKEND/src/main/java/com/neha/VideoStreamApp/services/CommentService.java
package com.neha.VideoStreamApp.services;

import com.neha.VideoStreamApp.dtos.response.CommentDto;

import java.util.List;

public interface CommentService {

    // User video par comment post kare (parentCommentId null = naya comment, set = reply)
    CommentDto addComment(String videoId, String text, String userEmail, String parentCommentId);

    // Ek video ke top-level comments (naye sabse pehle), har ek ke saath uske replies nested
    List<CommentDto> getCommentsByVideo(String videoId);

    // Comment delete kare — owner ya admin
    // Return: true = deleted, false = comment ka owner na hote hue bhi delete try kiya (caller 403 de)
    boolean deleteComment(String commentId, String requesterEmail, boolean isAdmin);

    // ── ADMIN ─────────────────────────────────────────────────────────────
    // Poore platform ke saare comments (moderation ke liye)
    List<CommentDto> getAllComments();

    // Admin — kisi bhi comment ko force delete kare
    void adminDeleteComment(String commentId);
}
